src/transactions/cdp.js

```diff
@@ -16,748 +30,865 @@ const MINRATIO = 140;
-	});
-	let txn4 = algosdk.makeAssetTransferTxnWithSuggestedParamsFromObject({
-		from: reserve.address,
-		to: info.address,
-		amount: microOpeningGard,
-		suggestedParams: params,
-		assetIndex: gardID
-	});
-	let r2_txns = [txn1, txn2, txn3, txn4]
-	algosdk.assignGroupID(r2_txns);
-	// r2_txns.splice(3, 1)
-	let txns = r1_txns.concat(r2_txns)
-	let stxns = await signGroup(info, txns)
-	console.log(stxns)
-	
-	// TODO: Adjust indexing based on including all txns
-
-	let r1_stxns = [stxns[0].blob, stxn2.blob];
-	// txn3
-	let start = 0;
-	if ( !optedIn ) {
-		r1_stxns.push(stxns[2].blob)
-		start = 1;
-	}
-	if ((await getCurrentUnix()) - start_time > 30) {
-		return {
-			alert: true,
-			text: "Aborted: Transaction review and signing took too long. Please try again."
-		}
-	}
-
-	const sendTxn1Promise = sendTxn(r1_stxns)
-
-	lsig = algosdk.makeLogicSig(reserve.logic, [algosdk.encodeUint64(1)]);
-	let stxn4 = algosdk.signLogicSigTransactionObject(txn4, lsig);
-		
-	let stxns2 = [stxns[start + 2].blob, stxns[start + 3].blob, stxns[start + 4].blob, stxn4.blob]
-	let response = await sendTxn1Promise;
-	response = await sendTxn(stxns2, "Successfully opened a CDP with ID: " + accountID + ".")
-	updateCDP(info.address, accountID, openingMicroALGOs, microOpeningGard);
-	return response
-	// XXX: May want to do something else besides this, a promise? loading screen?
+export function createOptInTxn(params, info, assetID) {
+  params.fee = 1000;
+  console.log("creating opt in for ", assetID);
+  let txn = algosdk.makeAssetTransferTxnWithSuggestedParamsFromObject({
+    from: info.address,
+    to: info.address,
+    amount: 0,
+    suggestedParams: params,
+    assetIndex: assetID,
+  });
+  return txn;
 }
 
-export async function mint(accountID, newGARD) {
+export async function openCDP(openingALGOs, openingGARD) {
+  if (openingGARD < 1) {
+    alert(
+      "Opening GARD needs to be above 1.\n" +
+        "Your opening GARD is is: " +
+        openingGARD,
+    );
+    return null;
+  }
+
+  // Setting up promises
+  const infoPromise = accountInfo();
+  const microOpeningGard = microGARD(openingGARD);
+  const devFeesPromise = calcDevFeesCurrent(microOpeningGard, false);
+  const paramsPromise = getParams(2000);
+
+  // XXX: Could add a nice check here to make sure the ratio is acceptable
+
+  const openingMicroALGOs = parseInt(openingALGOs * 1000000);
+  const ratio = calcRatio(openingMicroALGOs, openingGARD);
+  if (ratio < MINRATIO) {
+    return {
+      alert: true,
+      text:
+        "Ratio needs to be above " +
+        MINRATIO +
+        "%.\n" +
+        "Your ratio is: " +
+        calcRatio(openingMicroALGOs, openingGARD, true),
+    };
+  }
+
+  // Part 1: Opting in, creating needed info, etc.
+
+  const info = await infoPromise;
+  const devFees = await devFeesPromise;
+  const accountIDPromise = findOpenID(info.address);
+
+  if (
+    307000 +
+      openingMicroALGOs +
+      100000 * (info["assets"].length + 1) +
```

Comment:
This error catch should change to +3 (for 3 potential additional opt ins), and the same goes for the Maximum Deposit alert below. Ideally, we should do this alert later. See my comments on what the "start" variable represents

Reply:
Maybe this should be +4 come to think of it. I think when this was initially added we forgot to account for opt-ins altogether,